lib/class-wp-theme-json-resolver-gutenberg.php

```diff
@@ -563,6 +563,21 @@ public static function get_user_data() {
 			}
 		}
 
+		/*
+		 * The Separator block uses different CSS properties for its color depending
+		 * on how it is being rendered e.g. as "content" for the Dots style, or
+		 * as a border etc.
+		 * 
+		 * Uses are only presented with a single color control for background. Any
+		 * selection of a background color should be applied to the other paths
+		 * so it can be honoured.
+		 */
+		$separator_color =  $config['styles']['blocks']['core/separator']['color']['background'] ?? null;
+		if ( $separator_color ) {
+			_wp_array_set( $config, array( 'styles', 'blocks', 'core/separator', 'color', 'text' ), $separator_color );
+			_wp_array_set( $config, array( 'styles', 'blocks', 'core/separator', 'border', 'color' ), $separator_color );
+		}
```

Comment:
I like the idea of setting things in the user CPT. 

So, if a taking the example of TT4 and TT5 mentioned over in https://github.com/WordPress/gutenberg/issues/57297#issuecomment-2467042865, or any theme that specifies a border only and not a background, any updates to global styles background colors would overwrite it. 

In principle it seems more robust to do this operation at the source data (the merged object), so that any operation consuming this data wouldn't have to compute the styles all the time.

Speaking hypothetically, what would happen to the use of `update_separator_declarations` in `get_styles_for_block`?

I get the feeling this block will have some funky backwards compatibility issues later in life.

For example, throw in the hypothetical that the separator block gets padding support one day. The border color could different to the background color on purpose. 

I guess either way the existing `update_separator_declarations` would have to be updated to handle such a situation.

All in all, letting themes make their own mistakes, but forcing things at the global styles level appears to be a reasonable compromise.

Thanks for getting this discussion going @aaronrobertshaw 

Hopefully we get some more context from https://github.com/WordPress/gutenberg/pull/44943#issuecomment-2497000560

Reply:
> In principle it seems more robust to do this operation at the source data (the merged object), so that any operation consuming this data wouldn't have to compute the styles all the time.

I'm not sure I follow this comment 100%, sorry. 

I have a suspicion this is about filtering only in `get_merged_data`? I looked at that so that the copying of the background value into all the different paths wasn't saved within the Global Styles CPT but this approach would require instantiating additional `WP_Theme_JSON_Data_Gutenberg` instances which in turn instantiate more `WP_Theme_JSON_Gutenberg` instances causing a performance hit I wanted to avoid. This is why it was done after retrieving and decoding the CPT and before applying filters on the PHP side.

This PR still only filters the merged data on the JS side though as I was experimenting with still trying to avoid saving all the different path values and keep it to only the `background` color. Part of the next steps here is to consolidate everything including filtering at the same point on both JS and PHP sides. That might require moving the tweaks from `mergeBaseAndUserConfigs` to `useGlobalStylesUserConfig`.



> For example, throw in the hypothetical that the separator block gets padding support one day. The border color could different to the background color on purpose.

True but I suspect a Group block with no content would just be used instead as that's what the separator effectively becomes. I'd say it's a Spacer but that's been discussed on block support adoption PRs as being "layout-only".

> Speaking hypothetically, what would happen to the use of update_separator_declarations in get_styles_for_block?
> ...
> I guess either way the existing update_separator_declarations would have to be updated to handle such a situation.

The pre-existing hack was kept in this PR just to test the hypothesis that forcing the user selection in Global Styles wouldn't break existing block styles. 

My intention now the hypothesis seems to be proven is to consolidate the handling of the separator styles as much as possible. The declarations hack really seems out of place to me. That should have been a filter of the theme.json data all along in my view, which can be done at the theme origin level. 

So the next step for this PR is to get rid of `update_separator_declarations` from the theme.json style generation and tweak the theme origin's theme.json data instead. The existing hack in `updateConfigWithSeparator` and `useGlobalStylesOutputWithConfig` probably gets removed entirely as theme data would be filtered PHP side before it is passed to the editor via the block editor settings.

That's the plan at least. As the onion layers get peeled back, it might all change.

Reply:
> > In principle it seems more robust to do this operation at the source data (the merged object), so that any operation consuming this data wouldn't have to compute the styles all the time.

> 
> I'm not sure I follow this comment 100%, sorry.

It's long-winded way of saying I agree with your approach here 😄 and typed without re-reading, sorry. 

Overriding at the user data, rather than later in styles declarations, means that anyone grabbing the user data, say, from `WP_Theme_JSON_Resolver_Gutenberg::get_merged_data()` will benefit. 

Unless I'm completely missing the point, which could be the case.

> True but I suspect a Group block with no content would just be used instead as that's what the separator effectively becomes

I'm wondering if there's still semantic value in using `<hr />` element. Anyway, it's a moot point at this stage.

Reply:
> I don't believe hr elements support padding.

Good call! I didn't bother to check. Makes sense to me 👍🏻